### PR TITLE
Realm Puzzle App Example

### PR DESCRIPTION
**Pull request by TimOliver**
This PR introduces the publicly ready version of Realm Puzzle demo app.

simulator screen shot oct 25 2016 5 54 42 pm
## Features
- Rebranded for Realm's latest styling.
- Updated to use the public version of Realm Mobile Platform.
- Triple-tap with three fingers to reset the puzzle.
- A login screen that will automatically generate the IP address as well as remember the user credentials from the last time.

/cc @jpsim

### CODE REVIEW

File: examples/ios/objc/Puzzle/Views/RLMStartView.m [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+    
+    if (self.bottomCellBackgroundView == nil) {
+        self.bottomCellBackgroundView = [[UIImageView alloc] initWithImage:[RLMStartView cellBackgroundImageBottom:YES]];
+        self.bottomCellBackgroundView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
+    }
+    
+    if (self.hostNameField == nil) {
+        self.hostNameField = [self newTextField];
+        self.hostNameField.placeholder = @"localhost";
+        if (kLocalIPAddress.length > 0) {
```
Comment (jpsim) [inline examples/ios/objc/Puzzle/Views/RLMStartView.m]: When would this condition ever evaluate to false? Why is this check needed at all?

**Comment by jpsim** [comment]
Why does this PR change the object store commit?

File: examples/ios/objc/Puzzle/AppDelegate.m [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+#import <Realm/Realm.h>
+
+@interface AppDelegate ()
+
+@end
+
+@implementation AppDelegate
+
+- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions {
+    application.idleTimerDisabled = YES;
```
Comment (jpsim) [inline examples/ios/objc/Puzzle/AppDelegate.m]: should we do this for the Draw app too?

File: examples/ios/objc/Puzzle/Constants.h [hunk 3]
```diff
@@ -0,0 +1 @@
+static NSString * const kIPAddress = @"172.20.20.96";
```
Comment (jpsim) [inline examples/ios/objc/Puzzle/Constants.h]: Shouldn't this file be `.gitignore`'d?

File: examples/ios/objc/Puzzle/Controllers/RLMViewController.m [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+#import "RLMViewController.h"
+#import <Realm/Realm.h>
+#import <Realm/RLMRealmConfiguration+Sync.h>
+#import <Realm/RLMSyncConfiguration.h>
+#import "RLMPuzzle.h"
+#import "RLMPuzzlePiece.h"
+#import "RLMPuzzleView.h"
+#import "RLMStartView.h"
+
+static NSString *kRLMPuzzleUUID = @"7387fa06-dffa-4891-97e1-9e35c3898658";
```
Comment (jpsim) [inline examples/ios/objc/Puzzle/Controllers/RLMViewController.m]: There's only ever a single puzzle right? Might be conceptually simpler if this is just the empty string then, or even better if the property is an integer and the primary key value is always just 0.

File: examples/ios/objc/Puzzle/Controllers/RLMViewController.m [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+@interface RLMViewController () <RLMPuzzleViewDelegate, UIAlertViewDelegate>
+
+@property (nonatomic, strong) RLMStartView *startView;
+@property (nonatomic, strong) RLMPuzzleView *puzzleView;
+@property (nonatomic, strong) NSMutableArray *puzzlePieces;
+
+@property (nonatomic, strong) RLMNotificationToken *notificationToken;
+
+@property (nonatomic, strong) RLMResults *puzzles;
+@property (nonatomic, strong) NSString *currentPuzzleID;
```
Comment (jpsim) [inline examples/ios/objc/Puzzle/Controllers/RLMViewController.m]: Looks like `currentPuzzleID` is only ever set to `kRLMPuzzleUUID`. I don't think this property is actually needed.

File: examples/ios/objc/Puzzle/Controllers/RLMViewController.m [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+
+- (void)setupNotifications;
```
Comment (jpsim) [inline examples/ios/objc/Puzzle/Controllers/RLMViewController.m]: Looks like all the method declarations are only used from within this file, and therefore don't need to be declared other than their implementation.

File: examples/ios/objc/Puzzle/Controllers/RLMViewController.m [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+- (void)updatePuzzleState;
+
+- (void)resetGestureRecognized:(UIGestureRecognizer *)gestureRecognizer;
+
+@end
+
+@implementation RLMViewController
+
+#pragma mark - Controller Lifecycle -
+- (instancetype)init
```
Comment (jpsim) [inline examples/ios/objc/Puzzle/Controllers/RLMViewController.m]: This `init` method just calls `super`. Why not remove?

File: examples/ios/objc/Puzzle/Controllers/RLMViewController.m [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+{
+    if (self = [super init]) {
+       
+    }
+    
+    return self;
+}
+
+- (void)dealloc
+{
```
Comment (jpsim) [inline examples/ios/objc/Puzzle/Controllers/RLMViewController.m]: the code in this PR uses a mix of opening braces on the same line and new line for methods. Please try to consistently follow the same style, ideally the ObjC code style used by Realm overall: https://github.com/realm/realm-cocoa/wiki/Objective-C-Style-Guide

File: examples/ios/objc/Puzzle/Controllers/RLMViewController.m [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+    [self removeNotifications];
+}
+
+- (void)viewDidLoad {
+    [super viewDidLoad];
+    
+    self.view.backgroundColor = [UIColor colorWithWhite:1.0f alpha:1.0f];
```
Comment (jpsim) [inline examples/ios/objc/Puzzle/Controllers/RLMViewController.m]: `[UIColor whiteColor];`?

File: examples/ios/objc/Puzzle/Controllers/RLMViewController.m [hunk 10]
```diff
@@ -0,0 +1,249 @@
...
+- (void)startPuzzle
+{
+    //Create the over-arching puzzle object
+    RLMPuzzle *newPuzzle = [RLMPuzzle objectForPrimaryKey:kRLMPuzzleUUID];
+    self.currentPuzzleID = newPuzzle.uuid;
+    
+    BOOL firstTime = NO;
+    if (newPuzzle.pieces.count == 0) {
+        //Create a data point for each puzzle piece
+        NSMutableArray *puzzlePieces = [NSMutableArray array];
```
Comment (jpsim) [inline examples/ios/objc/Puzzle/Controllers/RLMViewController.m]: Why not merge the code to create the initial puzzle pieces with the code that creates the initial puzzle?

Comment (jpsim) [inline examples/ios/objc/Puzzle/Controllers/RLMViewController.m]: Also, why are you populating the pieces in a mutable array, only to then add them all to the `pieces` RLMArray? Just add them directly to the `pieces` property?

File: examples/ios/objc/Puzzle/Controllers/RLMViewController.m [hunk 11]
```diff
@@ -0,0 +1,249 @@
...
+        [weakSelf updatePuzzleState];
+    };
+    
+    _notificationToken = [[RLMRealm defaultRealm] addNotificationBlock:block];
+}
+
+- (void)removeNotifications
+{
+    [self.notificationToken stop];
+    self.notificationToken = nil;
```
Comment (jpsim) [inline examples/ios/objc/Puzzle/Controllers/RLMViewController.m]: there's no need to `nil` out the notification token.

File: examples/ios/objc/Puzzle/Models/RLMPuzzlePiece.h [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+#import <Foundation/Foundation.h>
+#import <UIKit/UIKit.h>
+#import <Realm/Realm.h>
+
+@interface RLMPuzzlePiece : RLMObject
+
+@property NSInteger identifier;
+@property double x;
+@property double y;
+@property NSString *userID;
```
Comment (jpsim) [inline examples/ios/objc/Puzzle/Models/RLMPuzzlePiece.h]: What's this `userID` property used for? It doesn't seem to ever be read from.

File: examples/ios/objc/Puzzle/Models/RLMPuzzle.h [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+//
+////////////////////////////////////////////////////////////////////////////
+
+#import "RLMPuzzlePiece.h"
+
+@interface RLMPuzzle : RLMObject
+
+@property NSString *uuid;
+@property NSString *name;
+@property NSString *userID;
```
Comment (jpsim) [inline examples/ios/objc/Puzzle/Models/RLMPuzzle.h]: What's this `userID` property used for? It doesn't seem to ever be read from.

File: examples/ios/objc/Puzzle/Models/RLMPuzzle.h [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+@property NSDate *creationTime;
```
Comment (jpsim) [inline examples/ios/objc/Puzzle/Models/RLMPuzzle.h]: What's this `creationTime` property used for? It doesn't seem to ever be read from.

File: examples/ios/objc/Puzzle/Models/RLMPuzzle.h [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#import "RLMPuzzlePiece.h"
+
+@interface RLMPuzzle : RLMObject
+
+@property NSString *uuid;
+@property NSString *name;
```
Comment (jpsim) [inline examples/ios/objc/Puzzle/Models/RLMPuzzle.h]: if there's only ever just one puzzle created, and it always has the same name of "My Puzzle", there's really no reason to store that in the object, right?

File: examples/ios/objc/Puzzle/Views/RLMPuzzleView.m [hunk 16]
```diff
@@ -0,0 +1,212 @@
...
+@interface RLMPuzzleView ()
+
+@property (nonatomic, strong) NSArray *puzzlePieces;
+@property (nonatomic, strong) NSString *pieceNames; //String versions of the piece names
+@property (nonatomic, strong) NSDictionary *pieceDefaultLocations; //The CGPoints for each piece in default space
+@property (nonatomic, assign) CGPoint gestureOrigin;
+@property (nonatomic, assign) CGPoint gesturePieceOrigin;
+@property (nonatomic, weak)   UIView *draggingView;
+
+- (void)setupPuzzlePieces;
```
Comment (jpsim) [inline examples/ios/objc/Puzzle/Views/RLMPuzzleView.m]: same with these declarations, I don't think they're used from outside the file, so why forward-declare them here?

File: examples/ios/objc/Puzzle/Views/RLMStartView.m [hunk 17]
```diff
@@ -0,0 +1,335 @@
...
+    
+    if (self.hostNameField == nil) {
+        self.hostNameField = [self newTextField];
+        self.hostNameField.placeholder = @"localhost";
+        
+        NSString *hostName = [[NSUserDefaults standardUserDefaults] stringForKey:kRLMPuzzleHostNameKey];
+        if (hostName.length > 0) {
+            self.hostNameField.text = hostName;
+        }
+        else if (kLocalIPAddress.length > 0) {
```
Comment (jpsim) [inline examples/ios/objc/Puzzle/Views/RLMStartView.m]: The constant generated by the Xcode build phase is actually named `kIPAddress`...

File: examples/ios/objc/Puzzle/AppDelegate.m (hunk 2 above)
Comment (TimOliver) [inline examples/ios/objc/Puzzle/AppDelegate.m]: That's a good idea! Just filed a PR: https://github.com/realm/realm-cocoa/pull/4440

**Comment by jpsim** [comment]
@TimOliver are you planning on reviving this at some point?

**Comment by TimOliver** [comment]
Oh! I thought this had already been merged.

Hmm, at this point, it might be better to remove it and put it in the Realm Demos repo.

**Comment by jpsim** [comment]
If we do that hopefully you can still incorporate the changes you made here.

**Comment by jpsim** [comment]
@TimOliver do you want to keep this open or just port it over to the [realm-demos org](https://github.com/realm-demos)?

**Comment by TimOliver** [comment]
@jpsim I plan to port it over to realm-demos when I get the chance. But I just did a merge tonight so I could deploy it to our demo devices for tomorrow night. :)

**Comment by jpsim** [comment]
Any reason to keep this PR open then?

**Comment by TimOliver** [comment]
I've moved the project across to the Realm demos repo, so we can close it now. :)

**Comment by jpsim** [comment]
Please close this branch when you're done with it.

**Pull request closed without merging**